src/cryptenroll/cryptenroll-pkcs11.c

```diff
@@ -40,24 +40,78 @@ int enroll_pkcs11(
         if (!pkey)
                 return log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to extract public key from X.509 certificate.");
 
-        r = rsa_pkey_to_suitable_key_size(pkey, &decrypted_key_size);
-        if (r < 0)
-                return log_error_errno(r, "Failed to determine RSA public key size.");
-
-        log_debug("Generating %zu bytes random key.", decrypted_key_size);
-
-        decrypted_key = malloc(decrypted_key_size);
-        if (!decrypted_key)
-                return log_oom();
-
-        r = genuine_random_bytes(decrypted_key, decrypted_key_size, RANDOM_BLOCK);
-        if (r < 0)
-                return log_error_errno(r, "Failed to generate random key: %m");
-
-        r = rsa_encrypt_bytes(pkey, decrypted_key, decrypted_key_size, &encrypted_key, &encrypted_key_size);
-        if (r < 0)
-                return log_error_errno(r, "Failed to encrypt key: %m");
-
+        int type = EVP_PKEY_base_id(pkey);
+        if (type == EVP_PKEY_RSA) {
+                r = rsa_pkey_to_suitable_key_size(pkey, &decrypted_key_size);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to determine RSA public key size.");
+
+                log_debug("Generating %zu bytes random key.", decrypted_key_size);
+
+                decrypted_key = malloc(decrypted_key_size);
+                if (!decrypted_key)
+                        return log_oom();
+
+                r = genuine_random_bytes(decrypted_key, decrypted_key_size, RANDOM_BLOCK);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to generate random key: %m");
+
+                r = rsa_encrypt_bytes(pkey, decrypted_key, decrypted_key_size, &encrypted_key, &encrypted_key_size);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to encrypt key: %m");
+        }
+        else if (type == EVP_PKEY_EC) {
+                _cleanup_(EVP_PKEY_freep) EVP_PKEY *pkey_new = NULL;
+                _cleanup_(erase_and_freep) uint8_t *shared_secret = NULL;
+                size_t shared_secret_len = 0;
+                EC_KEY *ec_key = EVP_PKEY_get0_EC_KEY(pkey);
+                int nid;
+
+                if (!ec_key)
+                        return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "pkey doesn't have EC_KEY associated");
+
+                nid = EC_GROUP_get_curve_name(EC_KEY_get0_group(ec_key));
+                r = pkey_generate_ec_key(nid, &pkey_new);
+                if (r != 0)
+                        return log_error_errno(r, "Failed to generate ec key");
+
+                r = pkey_ecdh_derive_shared_secret(pkey_new, pkey, NULL, &shared_secret_len);
+                if (r != 0 || shared_secret_len == 0)
+                        return log_error_errno(r, "Failed to determine derived shared secret size");
```

Comment:
if r == 0 and shared_secret_len == 0 then you will pass a 0 as first parameter to log_error_errno(). This will result in an assertion actually, log_error_errno() expects an error to be passed, and 0 is not an error...

Can shared_secret_len == 0 actually happen? Reading the function it dosn't look like it. Any reason you explicitly check for this?

also, please suffix with `…: %m"` again.